### DISCUSSION TITLE
CarPlay idrive?

### DISCUSSION THREAD

**Discussion by Kdp89** [BimmerGestalt/AAIdrive#652] [Ideas]
I just stumbled upon this app and my congratulations: it’s amazing. Thank you for helping extend the functionality of older BMWs.

I know you have rejected the idea before: but could you consider doing a similar application for iOS?  I can think of some strong reasons for doing this:

1. highly relevant for id4 and id4evo cars: they don’t have access to car play.
2. Ability to use usb for audio which is missing in Aaidrive; so better sound quality.
3. Possibly less Complicated to execute (unless Apple adds it’s own restrictions)

I would be happy to contribute financially to the development and can dedicate an ios device for beta testing :)
For me the limited set that would delight is :- music player access: for CarPlay compatible apps and the ability to send navigation intents to the vehicle nav.

Thanks again for all the work done.

**Comment by hufman** [comment]
That would be nifty!
However, I don't know how to develop for iOS, and don't use iOS personally. Most of the functionality of AAIdrive is the integration with other apps on the phone (music apps, system notifications, system calendar), which would all need to be rewritten for a different OS. My impression of Apple is that apps are strictly forbidden from sharing data with each other.

The first step would be reverse engineering how BMW car apps on iOS even talk to the car: Is it the same TCP server proxy that is used on Android? How do they know what port, and what is the audio context ID, which Android provides via Broadcast Intent? How do they get the security challenge response, which Android apps can fetch from the BMW App?

AAIdrive's source code would be a good example to copy from, for some of the car-specific behavior. The actual car connection and the integrations to other Android systems would not carry over. I can of course provide support and try to answer questions if someone wanted to try building an iOS app, but the first layer of reverse engineering must be done first by someone more skilled than me.

**Reply by Kdp89** [reply]
"but the first layer of reverse engineering must be done first by someone more skilled than me."

Alas, that is regrettably not me.  And its interesting enough to dive into, solo, but I fear with spare time and learning curve issues, by the time I reach some semblance of understanding, by it would be well past the sell by date of my id4 car :-)

i am wondering if it maybe easier to get USB audio working on android 8+. the music app that most interests me (Roon ARC) does not work below android 9 and most phones that run android 7 natively, are painfully slow. once you update them to the latest apps.

**Comment by LACNR** [comment]
From my knowledge BMW had a good support and compatibilty with apple gadgets, whenever my girl connect her iPhone via e.g. USB serveral Apps Spotify, Apple Music, Deezer and some more are selectable on the media and connected drive listing without her having a BMW App. I could only imagine some iOS Version could someday remove that compatibility.

**Reply by hufman** [reply]
Whoa really? Would you mind sharing a video about this? Apple Music definitely was never part of BMW Apps, so I'm very interested in how this works! Do they show up as apps in the Media menu? I know Android's Bluetooth stack is meant to provide the apps in a Browse menu within the Bluetooth Music app, but I haven't seen anything similar over USB.

**Reply by LACNR** [reply]
Maybe I'm confusing Apple Music, since audible and other apps were showing up, you have or had to choose show more Apps and it will give you a List of other compatible apps in the iDrive Menu.
Yeah, if I rememeber correctly they appear as apps below the external devices text in the Multimedia tab.. yeah of course I'll keep in mind making a video about it 🙂

I forgot to say, when connecting an iPhone the iDrive automatically loads the app settings unlike for Android which usually required an Connected App. The only, for the app to show up you need to run it which you usually already did upon connecting. 
I don't know if it a special protocol with iOS or if the lighting cable is the key, but yeah.
